src/libsyntax/parse/parser.rs

```diff
@@ -3616,9 +3631,42 @@ impl<'a> Parser<'a> {
             }
         };
 
-        if self.expr_is_complete(&lhs) {
-            // Semi-statement forms are odd. See https://github.com/rust-lang/rust/issues/29071
-            return Ok(lhs);
+        match (self.expr_is_complete(&lhs), AssocOp::from_token(&self.token)) {
+            (true, None) => {
+                // Semi-statement forms are odd. See https://github.com/rust-lang/rust/issues/29071
+                return Ok(lhs);
+            }
+            (false, _) => {} // continue parsing the expression
+            (true, Some(AssocOp::Multiply)) | // `{ 42 } *foo = bar;`
```

Comment:
Presumably these aren't exhaustive? It seems a little ad hoc to match on a specific set of operators like this. Is there a more general way we could detect this?

Reply:
The exhaustive list is checked in [line 3648](https://github.com/rust-lang/rust/pull/60188/files#diff-da9d34ca1d0f6beee2838cf02e07345cR3648). These are checked first because they are ambiguous but also very common source of trouble, so we want to keep their info around in `abiguous_block_expr_parse` to improve diagnostics in the appropriate case in later stages. I can move these three to another method like `can_continue_expr_unambiguously` if you wish.

Reply:
Maybe just a comment explaining this above?

Reply:
done